core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -576,7 +579,12 @@ public int compare(ShardRouting o1,
                     if ((indexCmp = o1.index().compareTo(o2.index())) == 0) {
                         return o1.getId() - o2.getId();
                     }
-                    return indexCmp;
+                    // this comparator is more expensive than all the others up there
+                    // that's why it's added last even though it could be easier to read
+                    // if we'd apply it earlier. this comparator will only differentiate across
+                    // indices all shared of the same index is treated equally.
```

Comment:
hmm, is there a typo somewhere in the end (s/shared/shards maybe?), I don't understand it. But otherwise, I would have been happy with just "Apply secondaryComparator last as it is more expensive than other comparators"